### PR TITLE
Modify rule S6544: include cases where promises are not resolved or error handled

### PR DESCRIPTION
**Pull request by ilia-kebets-sonarsource**

### CODE REVIEW

**Review by yassin-kammoun-sonarsource — CHANGES_REQUESTED** [review]
It's getting difficult to understand what the rule is all about now. This is why I think we need to introduce subsections to structure a bit the rule description.

I would move the main description in the section "Why is this an issue" and rework it as follows:

_[start]_

Promises need to be resolved or awaited to perform the expected asynchronous task and return a value, if any. Otherwise, they are nothing else than a regular object.

### Unresolved promises

Forgetting to await a promise is a frequent mistake. There are places where it makes no sense to use a promise object as it is. This is likely a bug as the developer forgot to resolve it.

The rule forbids returning promises where another type is expected such as in:

* conditionals
* void returns
* spread operators

### Missing error handling

When a promise raises an exception but doesn't have proper error handling, the program's call stack will be inconsistent and difficult to understand. (would be nice to have an example with a misleading call stack or strack trace)

_[end]_

File: rules/S6544/javascript/rule.adoc [hunk 1]
```diff
@@ -12,23 +16,42 @@ This rule forbids returning promises where another type is expected such as in:
...
 The executor function of a promise can also be an async function. However, this usually denotes a mistake:
 
-- If an async executor function throws an error, the error won't cause the created promise to reject and will be lost. Therefore, this could make it difficult to debug and handle runtime errors.
-- If a promise executor function is using `await`, this means that it's not necessary to use the `Promise` constructor, or the scope of the Promise constructor can be reduced.
+* If an async executor function throws an error, the error won't cause the created promise to reject and will be lost. Therefore, this could make it difficult to debug and handle runtime errors.
+* If a promise executor function is using `await`, this means that it's not necessary to use the `Promise` constructor, or the scope of the Promise constructor can be reduced.
+
+==== Exceptions
+
+This rule can be ignored for promises that you know will always resolve like timers.
```
[the remark this answers is not in the document]
Comment (ilia-kebets-sonarsource) [reply rules/S6544/javascript/rule.adoc:36]: Timers are an example, but there are indeed cases when people promisify event based code, which cannot reject.

[the remark this answers is not in the document]
Comment (ilia-kebets-sonarsource) [reply rules/S6544/javascript/rule.adoc:36]: So it's not an implemented exception.

File: rules/S6544/javascript/rule.adoc [hunk 2]
```diff
@@ -12,23 +16,42 @@ This rule forbids returning promises where another type is expected such as in:
 
 Forgetting to await a promise is a frequent mistake. There are places where it makes no sense to use a promise, but the developer forgets to resolve it.
 
+Promises that aren't resolved will run at an unexpected time, when they do not contain error handling and they raise an exception, the associated call stack will be wrong.
```
[the remark this answers is not in the document]
Comment (ilia-kebets-sonarsource) [reply rules/S6544/javascript/rule.adoc:19]: hmm, I can improve this part. Indeed a promise will start when executed synchronously. But usually promisified code has some process.nextTick() or event-based logic which will run when your code gets on top of the event queue again.

File: rules/S6544/javascript/rule.adoc [hunk 3]
```diff
@@ -12,23 +16,42 @@ This rule forbids returning promises where another type is expected such as in:
...
+[source,javascript]
+----
+await new Promise(resolve => time.setTimeout(1000));
+----
 
 == How to fix it
 
-If you mistook the promise with its resolved value, await the promise so it gets resolved correctly.
-Otherwise, you might need to adapt your code accordingly.
+If you mistook the promise with its resolved value, await or resolve the promise so it gets resolved correctly, using an IIFE if necessary:
```
[the remark this answers is not in the document]
Comment (ilia-kebets-sonarsource) [reply rules/S6544/javascript/rule.adoc]: done

**Comment by ilia-kebets-sonarsource** [comment]
@yassin-kammoun-sonarsource for the sections "Unresolved promises" and "Missing error handling", it is currently not allowed to have subsections for "why is this an issue", other than "potential impact", that's why I used underlines.

**Review by yassin-kammoun-sonarsource — APPROVED** [review]

File: rules/S6544/javascript/rule.adoc [hunk 4]
```diff
@@ -1,34 +1,54 @@
...
+[source,javascript]
+----
+await new Promise(resolve => time.setTimeout(1000));
+----
 
 == How to fix it
 
-If you mistook the promise with its resolved value, await the promise so it gets resolved correctly.
-Otherwise, you might need to adapt your code accordingly.
+If you mistook the promise with its resolved value, await or resolve the promise so it gets resolved correctly, using an "immediately invoked function expression" (IIFE) if necessary:
```
Comment (cynthiabethea) [inline rules/S6544/javascript/rule.adoc:42]: picky suggestion, I would reword this as:

If you mistakenly treated a promise as its resolved value, you can ensure it is properly resolved by using await or resolve on the promise. In some cases, you may need to use an "immediately invoked function expression" (IIFE).

**Review by cynthiabethea — APPROVED** [review]
added a few suggestions to consider
